Develop > Development Pathways (#161)
polkadot-developers/polkadot-docs: develop/development-pathways/index.md

---
title: Development Pathways
description: Explore the development pathways in the Polkadot ecosystem, from building dApps to creating core blockchain functionality and customizable runtimes.
---

# Development Pathways

## Introduction

> The Polkadot ecosystem provides two main development pathways:

Reviewer: I think it would be more accurate to say: The Polkadot SDK... instead of "polkadot ecosystem"

```mermaid
graph TD
    A[Development Pathways]
    A --> B[Client-side Development]
    A --> C[Protocol Development]
    C --> D[Smart Contracts]
    C --> E[Runtimes]
```

## Client-side Development

The client-side development path is dedicated to building applications that interact with Polkadot SDK-based blockchains and enhance user engagement with the network. While decentralized applications (dApps) are a significant focus, this pathway also includes developing other tools and interfaces that expand users' interactions with blockchain data and services.

Client-side developers can build:

- **Decentralized Applications (dApps)** - these applications leverage the blockchain's smart contracts or runtimes to offer a wide range of features, from financial services to gaming and social applications, all accessible directly by end-users

- **Command-Line Interfaces (CLIs)** - CLI tools empower developers and technical users to interact with the blockchain programmatically. These tools enable tasks like querying the blockchain, deploying smart contracts, managing wallets, and monitoring network status

> - **Data Analytics and Visualization Tools** - developers can create tools that aggregate, analyze, and visualize on-chain data to help users and businesses understand trends, track transactions, and gain insights into the network's health and usage

Reviewer: or even split them into 2 items:
- data analytics - and talk about big data analysis
- visualization or querying tools - like indexers and so on..

- **Wallets** - securely managing accounts and private keys is crucial for blockchain users. Client-side development includes building user-friendly wallets, account management tools, and extensions that integrate seamlessly with the ecosystem

- **Explorers and Dashboards** -  blockchain explorers allow users to view and search on-chain data, including blocks, transactions, and accounts. Dashboards provide a more interactive interface for users to monitor critical metrics, such as staking rewards, governance proposals, and network performance

> These applications can leverage the Polkadot blockchain's underlying protocol features to create robust, secure, and scalable solutions, allowing users to interact with the ecosystem innovatively. The Client-side Development pathway is ideal for developers interested in enhancing user experiences and building applications that bring the power of decentralized networks to a broader audience.

Reviewer: I'm not super convinced about this paragraph tbh

```suggestion
These applications can leverage the Polkadot blockchain's underlying protocol features to create robust, secure, and scalable solutions, allowing users to interact with the ecosystem. The Client-side development pathway is ideal for developers interested in enhancing user experiences and building applications that bring the power of decentralized networks to a broader audience.
```

Check the [API Libraries](TODO: update-path){target=\_blank} section for essential tools to interact with Polkadot SDK-based blockchain data and protocol features.

## Protocol Development

The protocol development path encompasses two key areas:

- Smart contracts

> - Runtimes

Reviewer: Add a short sentence for each one to briefly introduce the next sections

Author: I have reordered the section and linked smart contracts and runtimes to the subsections below, wdyt? I don't think we can explain smart contracts or runtimes with just a sentence

When developing a smart contract, you create a sandboxed program that executes specific logic associated with a chain address. Unlike pallets, which integrate directly into a blockchain's runtime and define its core logic, smart contracts operate independently in a more isolated environment. Pallets can manage critical tasks like block production or governance and can be added, modified, or removed through forkless runtime upgrades. Smart contracts, on the other hand, require careful upgrade planning, as their functionality can only be updated with explicit coding mechanisms for future changes. Parachains offer flexibility, allowing developers to define the environment where these contracts run and enabling others to build on it. These two options also differ in how they handle fees. Smart contracts follow a gas metering model for execution costs, while pallets and runtimes offer more flexible options for fee models.

!!!info "Additional information"
    - Refer to the [Runtime vs. Smart Contracts](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/runtime_vs_smart_contract/index.html){target=\_blank} section of the Polkadot SDK Rust docs

>     - Refer to this Stack Overflow post for a technically deeper discussion of when a developer might choose to develop a runtime versus a smart contract: [When should one build a Substrate (Polkadot SDK) runtime versus a Substrate (Polkadot SDK) smart contract?](https://stackoverflow.com/a/56041305){target=\_blank}

Reviewer: ```suggestion
    - Refer to the [When should one build a Substrate (Polkadot SDK) runtime versus a Substrate (Polkadot SDK) smart contract?](https://stackoverflow.com/a/56041305){target=\_blank} post on Stack Overflow for a technically deeper discussion of when a developer might choose to develop a runtime versus a smart contract
```

### Smart Contracts

> When building smart contracts, developers can define the specific logic and functionality that will be executed. This enables the creation of highly customized, decentralized applications that leverage the security and transparency of the underlying blockchain.

Reviewer: I think this is too generic, when building pallets devs can also do the same:
"When building smart contracts, developers can define the specific logic and functionality that will be executed."

Author: I have rewritten the section, pls check it again

Some key benefits of developing smart contracts include ease of development, faster time to market, and permissionless deployment. Smart contracts allow developers to quickly build and deploy decentralized applications without complex infrastructure or intermediaries. This accelerates the development lifecycle and enables rapid innovation within the Polkadot ecosystem.

> For more information on developing smart contracts in the Polkadot ecosystem, check the [Smart Contracts](TODO: update-path){target=\_blank} section.

Reviewer: here we should link to the smart contracts overview page

### Runtimes

Runtimes are the core building blocks that define the logic and functionality of Polkadot SDK-based blockchains. Developers can customize and extend the features of their blockchain, allowing for tighter integration with critical network tasks such as block production, consensus mechanisms, and governance processes.

Runtimes can be upgraded through forkless runtime updates, enabling seamless evolution of the blockchain without disrupting existing functionality.

> Developers can define the parameters, rules, and behaviors that shape their blockchain network. This includes token economics, transaction fees, permissions, and more. Using the Polkadot SDK, teams can rapidly iterate on their blockchain designs, experiment with new features, and deploy highly specialized networks tailored to their specific use cases.

Reviewer: ```suggestion
Developers can define the parameters, rules, and behaviors that shape their blockchain network. This includes token economics, transaction fees, permissions, and more. Using the Polkadot SDK, teams can iterate on their blockchain designs, experiment with new features, and deploy highly specialized networks tailored to their specific use cases.
```

> For those interested in delving deeper into runtime development, explore the dedicated [Blockchains](){target=\_blank} section.

Reviewer: However, I think we already have the structure to link this to the Custom Blockchain section in Develop

```suggestion
For those interested in delving deeper into runtime development, explore the dedicated [Blockchains](TODO: update path){target=\_blank} section.
```

Author: I've just linked the custom blockchains section

---
title: Development Pathways
description: Explore the development pathways in the Polkadot ecosystem, from building dApps to creating core blockchain functionality and customizable runtimes.
---

# Development Pathways

## Introduction

Developers can choose from different development pathways to build applications and core blockchain functionality. Each pathway caters to different types of projects and developer skill sets, while complementing one another within the broader network.

The Polkadot ecosystem provides two main development pathways:

- **Protocol development** - focuses on creating core components that define blockchain functionality, including developing runtime modules or implementing smart contracts to handle on-chain logic. This pathway is ideal for developers interested in contributing to the foundational layers of the Polkadot network or creating custom blockchain logic

> - **Client-side development** -  centers on building decentralized applications (dApps) that interact with the blockchain. Developers in this pathway create user-facing applications that leverage blockchain data and capabilities, enabling various use cases for end-users

Reviewer 2: ```suggestion
- **Client-side development** - centers on building decentralized applications (dApps) that interact with the blockchain. Developers in this pathway create user-facing applications that leverage blockchain data and capabilities, enabling various use cases for end-users
```

```mermaid
graph TD
    A[Development Pathways]
    A --> B[Client-side Development]
    A --> C[Protocol Development]
    C --> D[Smart Contracts]
    C --> E[Runtimes]
```

## Protocol Development

The protocol development path encompasses two key areas:

- [Smart contracts](#smart-contracts)
- [Runtimes](#runtimes)

> ### Smart Contracts

Reviewer 2: This feels kinda weird here. I don't know maybe it's just me. When I think a bout Protocol development, I think about  building a blockchain, creating custom runtimes, etc.

Smart contract as a part of protocol development makes me think ok how are blockchains enabling smart contract development. Not actually like developing smart contracts. So the second paragraph in particular doesn't make sense to me entirely here. But I don't know 🤷‍♀️

Author: @eshaben  I agree it feels weird. What do you think about having three sections instead of these 2?
Something like:
- Blockchain Development
- Smart Contract Development
- Client-Side Development

Reviewer 2: Yeah, sounds good to me!

Smart contracts are sandboxed programs that run within a virtual machine on the blockchain. These deterministic pieces of code are deployed at specific blockchain addresses and execute predefined logic when triggered by transactions. Because they run in an isolated environment, they provide enhanced security and predictable execution. Smart contracts can be deployed permissionlessly, allowing any developer to create and launch applications without requiring special access or permissions. They enable developers to create trustless applications by encoding rules, conditions, and state transitions that leverage the security and transparency of the underlying blockchain.

Some key benefits of developing smart contracts include ease of development, faster time to market, and permissionless deployment. Smart contracts allow developers to quickly build and deploy decentralized applications without complex infrastructure or intermediaries. This accelerates the development lifecycle and enables rapid innovation within the Polkadot ecosystem.

> For more information on developing smart contracts in the Polkadot ecosystem, check the [Smart Contracts](/develop/smart-contracts/overview){target=\_blank} section.

Reviewer 2: Yeah, point to this link is kind of solidifying my thoughts above that this section feels out of place. This link points to developers building smart contracts on top of a blockchain rather than developing a blockchain

!!!info "Additional information"
    - Refer to the [Runtime vs. Smart Contracts](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/runtime_vs_smart_contract/index.html){target=\_blank} section of the Polkadot SDK Rust docs
    - Refer to the [When should one build a Substrate (Polkadot SDK) runtime versus a Substrate (Polkadot SDK) smart contract?](https://stackoverflow.com/a/56041305){target=\_blank} post on Stack Overflow for a technically deeper discussion of when a developer might choose to develop a runtime versus a smart contract

> ## Client-side Development

Reviewer 2: ```suggestion
## Client-Side Development
```

> The client-side development path is dedicated to building applications that interact with Polkadot SDK-based blockchains and enhance user engagement with the network. While decentralized applications (dApps) are a significant focus, this pathway also includes developing other tools and interfaces that expand users' interactions with blockchain data and services.

Reviewer 2: This adds to my point above too: smart contracts interact with Polkadot SDK-based blockchains

- **Data Analytics and Visualization Tools** - developers can create tools that aggregate, analyze, and visualize on-chain data to help users and businesses understand trends, track transactions, and gain insights into the network's health and usage

- **Wallets** - securely managing accounts and private keys is crucial for blockchain users. Client-side development includes building user-friendly wallets, account management tools, and extensions that integrate seamlessly with the ecosystem

- **Explorers and Dashboards** -  blockchain explorers allow users to view and search on-chain data, including blocks, transactions, and accounts. Dashboards provide a more interactive interface for users to monitor critical metrics, such as staking rewards, governance proposals, and network performance

These applications can leverage the Polkadot blockchain's underlying protocol features to create solutions that allow users to interact with the ecosystem. The Client-side development pathway is ideal for developers interested in enhancing user experiences and building applications that bring the power of decentralized networks to a broader audience.

> Check the [API Libraries](TODO: update-path){target=\_blank} section for essential tools to interact with Polkadot SDK-based blockchain data and protocol features.

Reviewer: Maybe we can link to the index of the `api-libraries` section
```suggestion
Check the [API Libraries](/develop/toolkit/api-libraries/){target=\_blank} section for essential tools to interact with Polkadot SDK-based blockchain data and protocol features.
```
